src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -913,10 +913,20 @@ private async Task PopulateEventsCacheAsync(CancellationToken cancellationToken)
                 }
             }
 
-            static ImmutableArray<CompilationEvent> dequeueGeneratedCompilationEvents(AsyncQueue<CompilationEvent> eventQueue, ImmutableArray<AdditionalText> additionalFiles)
+            static ImmutableArray<CompilationEvent> dequeueGeneratedCompilationEvents(
+                AsyncQueue<CompilationEvent> eventQueue,
+                Compilation compilation,
+                AnalysisScope analysisScope,
+                ImmutableArray<AdditionalText> additionalFiles)
             {
                 var builder = ImmutableArray.CreateBuilder<CompilationEvent>();
 
+                // We synthesize a span-based CompilationUnitCompletedEvent for improved performance for computing semantic diagnostics
+                // of compiler diagnostic analyzer. See https://github.com/dotnet/roslyn/issues/56843 for more details.
+                var needsSpanBasedCompilationUnitCompletedEvent = analysisScope.FilterSpanOpt.HasValue &&
+                    analysisScope.IsSingleFileAnalysis &&
+                    !analysisScope.IsSyntacticSingleFileAnalysis;
```

Comment:
i don't quite understand the 'sytnactic single file analysis' distinction here.  wouldn't span-based make sense for that as well?

Reply:
Yes, it would, but the current public APIs for getting syntax diagnostics do not take in a filter span: https://github.com/dotnet/roslyn/blob/af66de2478e114033e25ade23f05d69968911236/src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs#L509-L535

Reply:
Also syntax diagnostics analysis is not driven by compilation events, so it wouldn't make sense in this specific code.